---
layout: default
title: Publication
---

### Publication

Feel free to email me <huazhou@ucla.edu> if you have any comments. 

D Kim, A Jensen, K Jones, S Raghavan, L Phillips, A Hung, Y Sun, G Li, P Reaven, **H Zhou**, and J Zhou. (2023) A platform for phenotyping disease progression and associated longitudinal risk factors in large-scale EHRs, with application to incident diabetes complications in the UK Biobank, [_JAMIA Open_](https://doi.org/10.1093/jamiaopen/ooad006), 6(1):ooad006. \[[pdf](./media/pdf/Kim23UKBPhenotyping.pdf)\]  

D Kim, A Binder, **H Zhou**, and S Jung. (2023) DNA methylation patterns associated with breast cancer prognosis that are specific to tumor subtype and menopausal status, [_Frontier in Genetics_](https://doi.org/10.3389/fgene.2023.1133443), 14:1133443. \[[pdf](./media/pdf/Kim23BCEwas.pdf)\]   

Q Heng, **H Zhou**, and E Chi. (2023) Bayesian trend filtering via proximal Markov Chain Monte Carlo, [_Journal of Computational and Graphical Statistics_](https://doi.org/10.1080/10618600.2023.2170089), in press. [arXiv](https://arxiv.org/abs/2201.00092)

S Ko, B Chu, D Peterson, C Okenwa, J Papp, D Alexander, E Sobel, **H Zhou**, and K Lange. (2023) Unsupervised discovery of ancestry informative markers and genetic admixture proportions in biobank-scale data sets, [_The American Journal of Human Genetics_](https://doi.org/10.1016/j.ajhg.2022.12.008), 110:314-325. \[[pdf](./media/pdf/Ko23OpenAdmixture.pdf)\]   

K Lange and **H Zhou**. (2022) A lagacy of EM algorithms. [_International Statistical Review_](http://dx.doi.org/10.1111/insr.12526), 92(S1): S52-S66. \[[pdf](./media/pdf/LangeZhou22NanLairdEM.pdf)\]  

J Won, T Zhang, and **H Zhou**. (2022) Orthogonal trace-sum maximization: tightness of the semidefinite relaxation and guarantee of locally optimal solutions. [_SIAM Journal of Optimization_](https://epubs.siam.org/doi/abs/10.1137/21M1422707) 32(3):2180-2207. \[[pdf](./media/pdf/WonZhangZhou22OTSM.pdf)\]  

A Landeros, O Madrid Padilla, **H Zhou**, and K Lange. (2022) Extensions to the proximal distance method of constrained optimization, [_Journal of Machine Learning Research_](http://jmlr.org/papers/v23/20-964.html), 23(182):1-45. \[[pdf](./media/pdf/LanderosPadillaZhouLange22ProxDist.pdf)\]  

S Ko, C German, A Jensen, J Shen, A Wang, D Mehrotra, Y Sun, J Sinsheimer, **H Zhou**, and J Zhou. (2022) GWAS of longitudinal trajectories at biobank scale, [_The American Journal of Human Genetics_](https://doi.org/10.1016/j.ajhg.2022.01.018), 109(3):433-445. \[[pdf](./media/pdf/Ko22TrajGWAS.pdf)\]

J Kim, A Jensen, S Ko, S Raghavan, L Philips, A Hung, Y Sun, **H Zhou**, P Reaven, and J Zhou. (2022) Systematic heritability and heritability enrichment analysis for diabetes complications in UK Biobank and ACCORD studies, [_Diabetes_](https://doi.org/10.2337/db21-0839), 71(5):1137-1148. \[[pdf](./media/pdf/Kim22DiabetesCompHeritability.pdf)\]

X Zhou, J Zhou, and **H Zhou**. (2022) Bag of little bootstraps for massive and distributed longitudinal data, [_Statistical Analysis and Data Mining_](https://doi.org/10.1002/sam.11563), 15(3):314-321. \[[pdf](./media/pdf/ZhouZhouZhou21LMM-BLB.pdf)\]   

K Doubleday, J Zhou, **H Zhou**, and H Fu. (2022) Risk controlled decision trees and random forests for precision medicine, [_Statistics in Medicine_](https://doi.org/10.1002/sim.9253), 41(4):719-735. \[[pdf](./media/pdf/DoubledayZhouZhouFu22ConstrRF.pdf)\]   

S Li, N Li, H Wang, J Zhou, **H Zhou**, and G Li. (2022) Efficient algorithms and implementation of a semiparametric joint model for longitudinal and competing risks data with applications to massive biobank data, [_Computational and Mathematical Methods in Medicine_](https://doi.org/10.1155/2022/1362913), 2022:1362913. \[[pdf](./media/pdf/LiLiWangZhouZhouLi22BiobankCmpRisk.pdf)\] 

B Chu, E Sobel, R Wasiolek, S Ko, J Sinsheimer, **H Zhou**, and K Lange. (2021) A fast data-driven method for genotype imputation, phasing, and local ancestry inference: MendelImpute.jl, [_Bioinformatics_](https://doi.org/10.1093/bioinformatics/btab489), 37(24):4756-4763. \[[pdf](./media/pdf/ChuSobelWasiolekKoSinsheimerZhouLange21MendelImpute.pdf)\]   

S Ko, **H Zhou**, J Zhou, and J Won. (2022) High-performance statistical computing in the computing environments of the 2020s. [_Statistical Science_](https://doi.org/10.1214/21-STS835), 37(4):494-518. \[[pdf](./media/pdf/KoZhouZhouWon21StatComp.pdf)\]   

K Lange, J Won, A Landeros, and **H Zhou**. (2021) Nonconvex optimization via MM algorithms: convergence theory. [_Wiley StatsRef: Statistics Reference Online_](https://doi.org/10.1002/9781118445112.stat08295). \[[pdf](./media/pdf/LangeWonAlfonsoZhou21MMConv.pdf)\]  

J Kim, J Shen, A  Wang, D Mehrotra, S Ko, J Zhou, and **H Zhou**. (2021) VCSEL: Prioritizing SNP-set by penalized variance component selection, [_Annals of Applied Statistics_](http://dx.doi.org/10.1214/21-AOAS1491), 15(4):1652-1672. \[[pdf](./media/pdf/KimShenWangMehrotraKoZhouZhou21VCSEL.pdf)\]  
Penalization methods for selecting variance components.  

C German, J Sinsheimer, J Zhou, and **H Zhou**. (2021) WiSER: robust and scalable estimation and inference of within-subject variances from intensive longitudinal data, [_Biometrics_](https://doi.org/10.1111/biom.13506), 78(4):1313-1327. \[[pdf](./media/pdf/GermanSinsheimerZhouZhou21WiSER.pdf)\]  
Efficient method and software for studying the mean and within-subject variability of longitudinal measurements.  

J Won, **H Zhou**, and K Lange. (2021) Orthogonal trace-sum maximization: applications, local algorithms, and global optimality. [_SIAM Journal on Matrix Analysis and Applications_](https://doi.org/10.1137/20M1363388), 42(2):859-882. \[[pdf](./media/pdf/WonZhouLange21OTSM.pdf)\]  
A simple certificate for checking when a local solution is global optimal for generalized canonical correlation analysis (CCA), generalized procrustes problem, and other problems.  

W Yu, **H Zhou**, Y Choi, J Goldin, P Teng, W Wong, M McNitt-Gray, M Brown, and G Kim (2022). Multi-scale, domain knowledge-guided attention + random forest: a two-stage deep learning-based multi-scale guided attention models to diagnose idiopathic pulmonary fibrosis from computed tomography images, [_Medical Physics_](http://doi.org/10.1002/mp.16053), 50(2):894-905. \[[pdf](./media/pdf/YuZhouChoiGoldinTengWongMcNittGrayBrownKim22IPF.pdf)\]  

W Yu, **H Zhou**, J Goldin, W Wong, and G Kim. (2021) End-to-end domain knowledge assisted automatic diagnosis of idiopathic pulmonary fibrosis (IPF) using high resolution computed tomography (HRCT), [_Medical Physics_](https://doi.org/10.1002/mp.14754), 48(5):2458--2467. \[[pdf](./media/pdf/YuZhouGoldinWongKim21IPF.pdf)\]   
Deep learning tool for diagnosing IPF from CT scans.  

S Ji, C German, K Lange, J Sinsheimer, **H Zhou**, J Zhou, and E Sobel. (2021) Modern simulation utilities for genetic analysis, [_BMC Bioinformatics_](https://doi.org/10.1186/s12859-021-04086-8), 22:228. \[[pdf](./media/pdf/Ji21GeneticSimulation.pdf)\]  

M Zhang, Y Liu, **H Zhou**, J Watkins, and J Zhou. (2021) A novel nonlinear dimension reduction approach to infer population structure for low-coverage sequencing data, [_BMC Bioinformatics_](https://doi.org/10.1186/s12859-021-04265-7), 22:348. \[[pdf](./media/pdf/ZhangLiuZhouWatkinsZhou21MCPCA.pdf)\] 

G Nam, ZF Zhang, JY Rao, **H Zhou**, and SY Jung. (2021) Interactions between adiponectin-pathway polymorphisms and obesity on postmenopausal breast cancer risk among African American women: the WHI SHARe Study, [_Frontiers in Oncology_](https://doi.org/10.3389/fonc.2021.698198), 11:698198. \[[pdf](./media/pdf/Nam21WHIShare.pdf)\]   

E Chi, B Gaines, W Sun, **H Zhou**, and J Yang. (2020) Provable convex co-clustering of tensors. [_Journal of Machine Learning Research_](https://jmlr.org/papers/v21/18-155.html), 21(214):1-58. \[[pdf](./media/pdf/ChiGainesSunZhouYang20CoCo.pdf)\]   
Convex co-clustering of tensor data.  

J Zhou, J Zhai, **H Zhou**, Y Chen, S Guerra, I Robey, G Weinstock, E Weinstock, Q Dong, K Knox, and H Twigg III. (2020) Supraglottic lung microbiome taxa are associated with pulmonary abnormalities in an HIV longitudinal cohort, [_American Journal of Respiratory and Critical Care Medicine_](https://www.atsjournals.org/doi/10.1164/rccm.202004-1086LE), 202(12):1727-1731. \[[pdf](./media/pdf/Zhou20LungMicrobiome.pdf)\]

I Ipsen, **H Zhou**. (2020) Probabilistic error analysis for inner products, [_SIAM Journal of Matrix Analysis_](https://epubs.siam.org/doi/abs/10.1137/19M1270434), 41(4):1726-1741. \[[pdf](./media/pdf/IpsenZhou20ProbError.pdf)\]  

J Day, **H Zhou**. (2020) OnlineStats.jl: A Julia package for statistics on data streams, [_Journal of Open Source Software_](https://doi.org/10.21105/joss.01816), 5(46):1816. \[[pdf](./media/pdf/DayZhou20OnlineStats.pdf)\]  
A highly efficient and extensible framework for performing statistical estimation and inference on streaming data.  

K Keys, **H Zhou**, K Lange. (2019) Proximal distance algorithms: theory and practice, [_Journal of Machine Learning Research_](http://jmlr.org/papers/v20/17-687.html), 20(66):1-38. \[[pdf](./media/pdf/KevinZhouLange19ProxDist.pdf)\]  
A class of algorithms for constrained optimization based on distance majorization.  

C German, J Sinsheimer, Y Klimentidis, **H Zhou**, and J Zhou. (2020) Ordered multinomial regression for genetic association analysis of ordinal phenotypes at Biobank scale, [_Genetic Epidemiology_](https://doi.org/10.1002/gepi.22276), 44:248-260. \[[pdf](./media/pdf/German19OrdinalGWAS.pdf)\]  
GWAS tool for ordinal traits.  

B Chu, K Keys, C German, **H Zhou**, J Sinsheimer, and K Lange. (2020) Iterative hard thresholding in GWAS: generalized linear models, prior weights, and double sparsity, [_GigaScience_](https://academic.oup.com/gigascience/article/9/6/giaa044/5850823), 9(6):gia044. [pdf](./media/pdf/Chu20IHTGWAS.pdf)  

B vonHoldt, A DeCandia, E Heppenheimer, H Janowitz-Koch, R Shi, **H Zhou**, C German, K Brzeski, K Cassidy, D Stahler, and J Sinsheimer. (2020) Heritability of inter‐pack aggression in a wild pedigreed population of North American gray wolves, [_Molecular Ecology_](https://doi.org/10.1111/mec.15349), 29(10):1764-1775. \[[pdf](./media/pdf/vonHoldt20GrayWolveAgression.pdf)\]  

W Hu, W Shen, **H Zhou**, and D Kong. (2020) Matrix linear discriminant analysis, [_Technometrics_](https://doi.org/10.1080/00401706.2019.1610069), 62(2):196-205. \[[pdf](./media/pdf/HuShenZhouKong20MatrixLDA.pdf)\]  
Linear discriminant analysis for matrix-variate data.  

**H Zhou**, L Hu, J Zhou, and K Lange. (2019) MM algorithms for variance components models. [_Journal of Computational and Graphical Statistics_](https://doi.org/10.1080/10618600.2018.1529601), 28(2):350-361. \[[pdf](./media/pdf/ZhouHuZhouLange19VCMM.pdf)\]  
A new optimization algorithm for find MLE of variance components models. 

**H Zhou**, J Sinsheimer, D Bates, B Chu, C German, S Ji, K Keys, J Kim, S Ko, G Mosher, J Papp, E Sobel, J Zhai, J Zhou, and K Lange. (2020) OpenMendel: a cooperative programming project for statistical genetics, [_Human Genetics_](https://doi.org/10.1007/s00439-019-02001-z), 139(1):61-71. \[[pdf](./media/pdf/Zhou19OpenMendel.pdf)\]   
Overview of the [OpenMendel project](https://github.com/openmendel).

K Doubleday, **H Zhou**, H Fu, and J Zhou. (2018) An algorithm for generating individualized treatment decision trees and random forests, [_Journal of Computational and Graphical Statistics_](https://doi.org/10.1080/10618600.2018.1451337), 27(4):849-860. \[[pdf](./media/pdf/DoubledayZhouFuZhou18IndividualizedTreatmentTrees.pdf)\]

B Gaines, J Kim, and **H Zhou**. (2018) Algorithms for fitting the constrained lasso, [_Journal of Computational and Graphical Statistics_](https://doi.org/10.1080/10618600.2018.1473777), 27(4):861-871. \[[pdf](./media/pdf/GainesKimZhou08CLasso.pdf)\]

L Hu, W Lu, J Zhou, and **H Zhou**. (2018) MM algorithms for variance component estimation and selection in logistic linear mixed models, [_Statistica Sinica_](https://doi.org/10.5705/ss.202017.0220), 28:1585-1605. \[[pdf](./media/pdf/HuLvZhouZhou19GLMM.pdf)\]  

EJ Min, E Chi, and **H Zhou**. (2020) Tensor canonical correlation analysis. [_STAT_](https://doi.org/10.1002/sta4.253), 8:e253. \[[pdf](./media/pdf/MinChiZhou20TCCA.pdf)\]   
Canonical correlation analysis (CCA) with tensor data.  

X Zhang, L Li, **H Zhou**, Y Zhou, D Shen, and ADNI. (2019) Tensor generalized estimating equations for longitudinal imaging analysis, [_Statistica Sinica_](https://doi.org/10.5705/ss.202017.0153), 29:1977-2005. \[[pdf](./media/pdf/Zhang19TensorGEE.pdf)\]   
GEE for tensor data.  

X Li, D Xu, **H Zhou**, and L Li. (2018) Tucker tensor regression and neuroimaging analysis, [_Statistics in Biosciences_](https://doi.org/10.1007/s12561-018-9215-6), 10(3):520-545. \[[pdf](./media/pdf/LiXuZhouLi18TuckerReg.pdf)\]  
Tucker version of the tensor regression.

J Zhai, J Kim, K Knox, H Twigg, **H Zhou**, and J Zhou. (2018) Variance component selection with applications to microbiome taxonomic data, [_Frontiers in Microbiology_](https://doi.org/10.3389/fmicb.2018.00509), 9:509. \[[pdf](./media/pdf/Zhai18MicrobiomeVCSelection.pdf)\]  
Variance component penalization method for microbiome taxa association study.

J Zhai, K Knox, H Twigg, **H Zhou**, and J Zhou. (2019) Exact variance component tests for longitudinal microbiome studies, [_Genetic Epidemiology_](https://doi.org/10.1002/gepi.22185), 43:250-262. \[[pdf](./media/pdf/ZhaiKnoxTwiggZhouZhou19MicrobiomeTest.pdf)\]

J Kim, Y Zhang, J Day, and **H Zhou**. (2018) MGLM: an R package for multivariate categorical data analysis. [_The R Journal_](https://journal.r-project.org/archive/2018/RJ-2018-015/index.html), 10(1):73-90. \[[pdf](./media/pdf/KimZhangDayZhou18MGLM.pdf)\]

C Li and **H Zhou**. (2017) svt: Singular value thresholding in MATLAB, [_Journal of Statistical Software_](https://www.jstatsoft.org/v081/c02), 81(2):1-13. \[[pdf](https://www.jstatsoft.org/index.php/jss/article/view/v081c02/v81c02.pdf)\]

J Zhou, T Hu, D Qiao, M Cho, and **H Zhou**. (2016) Boosting gene mapping power and efficiency with efficient exact variance component tests of SNP sets, [_Genetics_](http://www.genetics.org/content/204/3/921), 204(3):921-931. \[[pdf](./media/pdf/ZhouHuQiaoChoZhou16Boost.pdf)\]  
Exact score and likelihood ratio tests for testing SNP sets.  

Y Zhang, **H Zhou**, J Zhou, and W Sun. (2017) Regression models for multivariate count data. [_Journal of Computational and Graphical Statistics_](http://www.tandfonline.com/doi/abs/10.1080/10618600.2016.1154063), 26(1):1-13. \[[pdf](./media/pdf/ZhangZhouZhouSun17mglm.pdf)\]  
Regression model using Dirichlet-multinomial, negative multinomial, and generalized Dirichlet-multinomial distributions.  

**H Zhou**, J Blangero, T Dyer, K Chan, K Lange, and E Sobel. (2017) Fast genome-wide QTL association mapping
on pedigree and population data, [_Genetic Epidemiology_](http://onlinelibrary.wiley.com/doi/10.1002/gepi.21988/abstract), 41(3):174-186. \[[pdf](./media/pdf/ZhouBlangeroThomasChanLangeSobel17PedGWAS.pdf)\]  
Method for [Mendel software](https://www.genetics.ucla.edu/software/mendel) Option 29 (Pedigree GWAS).   

B Zhang, **H Zhou**, L Wang, and C Sung. (2017) Classification based on neuroimaging data by tensor boosting, [*International Joint Conference on Neural Networks (IJCNN)*](http://ieeexplore.ieee.org/abstract/document/7965985/), 1174-1179.  \[[pdf](./media/pdf/Zhang17TensorBoosting.pdf)\]  

**H Zhou**, J Zhou, T Hu, E Sobel, and K Lange. (2016) Genome-wide QTL and eQTL analyses using Mendel, [_BMC Proceedings_](http://rdcu.be/mHtr), 10(Suppl 7):10. \[[pdf](./media/pdf/Zhou16GAW19.pdf)\]  
QTL and eQTL analyses of the GAW19 data.  

N Zhao, J Chen, IM Carroll, T Ringel-Kulka, MP Epstein, **H Zhou**, J Zhou, Y Ringel, HZ Li, and MC Wu. (2015) Testing in microbiome profiling studies with the microbiome regression-based kernel association test (MiRKAT). [_The American Journal of Human Genetics_](http://www.sciencedirect.com/science/article/pii/S0002929715001408#), 96(5):797-807. \[[pdf](./media/pdf/Zhao15MiRKAT.pdf)\]  

W Sun, Y Liu, JJ Crowley, TH Chen, **H Zhou**, H Chu, S Huang, PF Kuan, Y Li, D Miller, G Shaw, Y Wu, V Zhabotynsky, L McMillan, F Zou, PF Sullivan and FPM de Villena. (2015) IsoDOT detects differential RNA-isoform expression/usage with respect to a categorical or continuous covariate with high sensitivity and specificity, [_Journal of American Statistical Association_](http://www.tandfonline.com/doi/abs/10.1080/01621459.2015.1040880), 110:975-986. \[[pdf](./media/pdf/Sun15IsoDOT.pdf)\]  

**H Zhou** and K Lange. (2015) Path following in the exact penalty method of convex programming, [_Computational Optimization and Applications_](http://link.springer.com/article/10.1007/s10589-015-9732-x), 61(3):609-634. \[[pdf](./media/pdf/ZhouLange15ConvProgPath.pdf)\]  

W Xiao, Y Wu and **H Zhou**. (2015) ConvexLAR: an extension of least angle regression, [_Journal of Computational and Graphical Statistics_](http://amstat.tandfonline.com/doi/full/10.1080/10618600.2014.962700), 24(3):603--626. \[[pdf](./media/pdf/XiaoWuZhou15ConvexLAR.pdf)\]  
Least angle regression (LAR) with a general convex loss.

**H Zhou** and Y Wu. (2014) A generic path algorithm for regularized statistical estimation, [_Journal of American Statistical Association_](http://www.tandfonline.com/doi/full/10.1080/01621459.2013.864166#.Up5KiGRDt4A), 109(506):686-699. \[[pdf](./media/pdf/ZhouWu14EPSODE.pdf)\]  
Path following for a convex loss plus a generalized lasso penalty.

E Chi, **H Zhou** and K Lange. (2014) Distance majorization and its applications, [_Mathematical Programming Series A_](http://link.springer.com/article/10.1007/s10107-013-0697-1), 146(1-2):409-436. \[[pdf](./media/pdf/ChiZhouLange14Distance.pdf)\]

K Lange, E Chi, and **H Zhou**. (2014) A brief survey of modern optimization for statisticians (with discussions by [Atchade and Michailidis](http://onlinelibrary.wiley.com/doi/10.1111/insr.12033/full), [Hunter](http://onlinelibrary.wiley.com/doi/10.1111/insr.12027/full), [Robert](http://onlinelibrary.wiley.com/doi/10.1111/insr.12040/full), and [rejoinder](http://onlinelibrary.wiley.com/doi/10.1111/insr.12030/full)), [_International Statistical Review_](http://onlinelibrary.wiley.com/doi/10.1111/insr.12022/abstract), 82(1):46-70. \[[pdf](./media/pdf/LangeChiZhou14OptmSurvey.pdf)\]

**H Zhou** and L Li. (2014) Regularized matrix regression, [_Journal of Royal Statistical Society Series B_](http://onlinelibrary.wiley.com/doi/10.1111/rssb.12031/abstract), 76(2):463-483. \[[pdf](./media/pdf/ZhouLi14RegMatrixReg.pdf)\]\[[software](./softwares/tensorreg)\]  
Soft-thresholding for regression with matrix covariates.

K Lange, JC Papp, JS Sinsheimer, R Sripracha, **H Zhou**, and E Sobel. (2013) Mendel: the Swiss army knife of genetic analysis programs, [_Bioinformatics_][4], 29(12):1568-1570. \[[pdf](./media/pdf/Lange13Mendel.pdf)\]\[[Mendel][6]\]  
Summary of the new version of the comprehensive genetic analysis software Mendel.

E Chi, G Allen, **H Zhou**, O Kohannim, K Lange, and P Thompson. (2013) Imaging genetics via sparse canonical correlation analysis, [_Biomedical Imaging (ISBI), 2013 IEEE 10th International Symposium on_](http://ieeexplore.ieee.org/xpl/login.jsp?reload=true&tp=&arnumber=6556581&url=http%3A%2F%2Fieeexplore.ieee.org%2Fxpls%2Fabs_all.jsp%3Farnumber%3D6556581), pp740-743. \[[pdf](./media/pdf/Chi13TensorCCA.pdf)]  
Sparse canonical correlation analysis (CCA) for the tensor data.

**H Zhou**, J Zhou, E Sobel, and K Lange. (2014) Fast genome-wide pedigree QTL analysis using Mendel, [_BMC Proceedings_](http://www.biomedcentral.com/1753-6561/8/S1/S93), 8(Suppl 1):S93. \[[pdf](./media/pdf/ZhouZhouSobelLange14GAW18.pdf)\]  
Multivariate QTL analysis of the [GAW18][7] data.

**H Zhou**, L Li, and H Zhu. (2013) Tensor regression with applications in neuroimaging data analysis, [_Journal of American Statistical Association_](http://www.tandfonline.com/doi/abs/10.1080/01621459.2013.776499#.Ud7Ytz7Xjbw), 108(502):540-552. \[[pdf](./media/pdf/ZhouLiZhu13CPReg.pdf)\]\[[software](./softwares/tensorreg)\]  
Traditional regression takes a vector of covariates. We consider regression that takes an array, aka tensor, of covariates, such as in neuroimaging studies.

E Chi, **H Zhou**, G Chen, D Ortega, and K Lange. (2013) Genotype imputation via matrix completion, [_Genome Research_](http://genome.cshlp.org/content/23/3/509), 23:509-518. \[[pdf](./media/pdf/ChiZhouChenLange13Impute.pdf)\]\[[Mendel Impute](http://www.genetics.ucla.edu/software/)\]  
We successfully applied matrix completion method to the difficult genotype imputation problem. Similar imputation accuracy is achieved in order of magnitude less time than current methods.

K Lange and **H Zhou**. (2014) MM algorithms for geometric and signomial programming, [_Mathematical Programming Series A_][8], 143(1-2):339-356. \[[pdf](./media/pdf/LangeZhou14GP.pdf)\]  
A simple algorithm for minimizing posynomials and signomials.

L Riley, **H Zhou**, K Lange, J Sinsheimer, and M Sehl. (2012) Determining duration of HER2-targeted therapy using stem cell extinction models, [_PLoS ONE_][10], 7(12):e46613. \[[pdf](./media/pdf/Riley12Extinction.pdf)\]  
Prediction of optimal duration of stem-cell target treatment based on a stochastic model.

H Zhu, L Li, and **H Zhou**. (2012) Nonlinear dimension reduction with Wright-Fisher kernel for genotype aggregation and association mapping, [_Bioinformatics_][12], 28:i375--i381. \[[pdf](./media/pdf/ZhuLiZhou12SDR-Seq.pdf)\]  
Nonlinear dimension reduction using Markov chain kernels.

**H Zhou** and K Lange. (2013) A path algorithm for constrained estimation, [_Journal of Computational and Graphical Statistics_](http://amstat.tandfonline.com/doi/full/10.1080/10618600.2012.681248), 22(2):261-283. \[[pdf](./media/pdf/ZhouLange13LSPath.pdf)\]   
Path following for a quadratic criterion plus a generalized lasso penalty.

**H Zhou** and Y Zhang. (2012) EM vs MM: a case study, [Computational Statistics & Data Analysis](http://www.sciencedirect.com/science/article/pii/S0167947312002174), 56:3909--3920. \[[pdf](./media/pdf/ZhouZhang12EMvsMM.pdf)\]  
A comparison of EM (expectation-maximization) and MM (minorization-maximization) algorithms in the case of MLE for the Dirichlet-Multinomial distribution.

M Sehl, **H Zhou**, J Sinsheimer, and K Lange. (2011) Extinction models for cancer stem cell therapy, [_Mathematical Biosciences_][14], 234(2):132-146. \[[pdf](./media/pdf/SehlZhouSinsheimerLange12StemCell.pdf)\]  
A Markov process model for the stem cells in cancer patients under treatment. A mix of techniques, e.g., extreme value theory, fast Fourier transform, and orthogonal polynomial expansions, are used.

**H Zhou**, D Alexander, and K Lange. (2011) A quasi-Newton acceleration for high-dimensional optimization algorithms, [_Statistics and Computing_][16], 21(2):261-273. \[[pdf](./media/pdf/ZhouAlexanderLange11QN.pdf)\]  
A new quasi-Newton acceleration scheme particularly suitable for EM/MM algorithms on high dimensional problems.

**H Zhou** and K Lange. (2011) A fast procedure for calculating importance weights in bootstrap sampling, [_Computational Statistics & Data Analysis_][18], 55(1):26-33. \[[pdf](./media/pdf/ZhouLange11ImportanceSampling.pdf)\]  
The use of importance sampling can dramatically reduce the variance of boostrap estimates. However nowadays the sample size is often huge, say at order of 10^6-10^9. Then the calculation of importance weights becomes nontrivial. Here we propose a fast procedure that scales well with sample size.

**H Zhou**, D Alexander, M Sehl, J Sinsheimer, E Sobel, and K Lange. (2011) Penalized regression for genome-wide association screening of sequence data, [_Pacific Symposium of Biocomputing_][20], 16:106-117. \[[pdf](./media/pdf/ZhouAlexanderSehlSobelLange11PBS.pdf)\]\[[Mendel][6]\]  
This companion paper introduces weights to calibrate the penalties and covers more details on the implementation of penalized regression in the statistical genetics analysis software [Mendel][6].

**H Zhou**, M Sehl, J Sinsheimer, and K Lange. (2010) Association screening of common and rare genetic variants by penalized regression, [_Bioinformatics_](10.1093/bioinformatics/btq448), 26(19):2357-2382. \[[pdf](./media/pdf/ZhouSehlSinsheimerSobelLange10RVLasso.pdf)\]  
Application of penalized linear and logistic regressions to genome-wide association studies (GWAS). A mixture of lasso and group penalties are used to select causal rare variants present in sequence data.

**H Zhou**, K Lange, and M Suchard. (2010) Graphics processing units and high-dimensional optimization, [_Statistical Science_][24], 25:311-324. \[[pdf](./media/pdf/ZhouSuchardLange10GPU.pdf)\]  
The marriage of MM principle and modern graphical processing unit (GPU) technology gives a boost to the classical EM/MM type algorithms widely used in statistics.

**H Zhou** and K Lange. (2010) On the bumpy road to the dominant mode, [_Scandinavian Journal of Statistics_][26], 37:612-631. \[[pdf](./media/pdf/ZhouLange10Annealing.pdf)\]  
We propose several variants of deterministic annealing for finding the dominant mode in maximum likelihood estimation with some classical statistical problems.

**H Zhou** and K Lange. (2010) MM algorithms for some discrete multivariate distributions, [_Journal of Computational and Graphical Statistics_][28], 19(3):656-665. \[[pdf](./media/pdf/ZhouLange10MM.pdf)\]  
We designed MM algorithms (a generalization of the EM algorithm) for maximum likelihood estimation of some multivariate distributions frequently occurring in applications. Specific examples include Dirichlet-Multinomial, generalized Dirichlet-Multinomial, negative multinomial, a distribution due to Neerchal and Morel, the Ewens and Pitman sampling distributions, and the zero-truncated and zero-inflated version of these distributions.

M Sehl, J Sinsheimer, **H Zhou**, and K Lange. (2009) Differential destruction of stem cells: implications for targeted cancer stem cell therapy, [_Cancer Research_][30], 69(24):9481-9489. \[[pdf](./media/pdf/SehlSinsheimerZhouLange09Extinction.pdf)\]  
Clinical implications of the cancer extinction model are summarized in this paper.

**H Zhou** and K Lange. (2009) Rating movies and rating the raters who rate them, [_The American Statistician_][32], 63(4):297–307. \[[pdf](./media/pdf/ZhouLange09MovieRating.pdf)\]  
If you ever heard of [Netflix grand prize][34], you might be interested in reading this short report. Instead of the prediction problem challenged by Netflix, we focus more on the modeling. Our simple model is able to identify quirky raters, supply a ranking of movies, and should be able to predict unseen ratings. We only fit the model to the [MovieLens][35] data set. Interested readers should try this on the (much larger) Netflix data set.

**H Zhou** and K Lange. (2009) Composition Markov chains of multinomial type, [_Advances in Applied Probability_](https://doi.org/10.1239/aap/1240319585), 41(1):270-291. \[[pdf](./media/pdf/ZhouLange09Multinomial.pdf)\]  
We describe a class of Markov chains that take a system of multivariate Krawtchouk polynomials constructed by [Robert Griffiths][38] as eigenfunctions.  

K Khare and **H Zhou**. (2009) Rates of convergence of some multivariate Markov chains with polynomial eigenfunctions, [_Annals of Applied Probability_](10.1214/08-AAP562), 19(2):737-777. \[[pdf](./media/pdf/KhareZhou09AAP.pdf)\]  
We obtaine the convergence rates of the multivariate versions of several classical Markov chains using spectral method. Specific examples include the multivariate Moran process in population genetics and its variants in community ecology, Dirichlet-Multinomial Gibbs sampler, generalizations of Ehrenfest chains, multivariate normal autoregressive processes, and so on.

**H Zhou** and K Dorman. (2005) A branching process model of drug resistant HIV, book chapter in [_Deterministic and stochastic models of AIDS epidemics and HIV infections with intervention_](http://www.amazon.com/dp/9812561390), 457-496, World Sci. Publ., Hackensack, NJ.   
We apply the numerical methods for continuous-time multi-type branching processes with immigration to study the development of drug-resistant HIV in vivo.

   [2]: http://www.stat.ncsu.edu/people/zhou/courses/st810/notes/SurveyPaper.pdf
   [3]: http://arxiv.org/abs/1204.3331
   [4]: http://bioinformatics.oxfordjournals.org/content/29/12/1568
   [5]: https://sites.google.com/site/hwachou/publication/Lange13Mendel.pdf?attredirects=0
   [6]: http://www.genetics.ucla.edu/software/
   [7]: http://www.gaworkshop.org/gaw18/index.html
   [8]: http://link.springer.com/article/10.1007%2Fs10107-012-0612-1?LI=false
   [9]: https://sites.google.com/site/hwachou/publication/LangeZhou12GP.pdf?attredirects=0
   [10]: http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0046613
   [11]: https://sites.google.com/site/hwachou/publication/Riley12Extinction.pdf?attredirects=0
   [12]: http://bioinformatics.oxfordjournals.org/content/28/18/i375.full
   [13]: https://sites.google.com/site/hwachou/publication/ZhuLiZhou12SDR-Seq.pdf?attredirects=0
   [14]: http://www.sciencedirect.com/science/article/pii/S0025556411001465
   [15]: https://sites.google.com/site/hwachou/publication/SehlZhouSinsheimerLange12StemCell.pdf?attredirects=0
   [16]: http://www.springerlink.com/content/41637373403m8878/
   [17]: https://sites.google.com/site/hwachou/publication/ZhouAlexanderLange11QN.pdf?attredirects=0
   [18]: http://dx.doi.org/10.1016/j.csda.2010.04.019
   [19]: https://sites.google.com/site/hwachou/publication/ZhouLange11ImportanceSampling.pdf?attredirects=0
   [20]: http://www.worldscientific.com/doi/abs/10.1142/9789814335058_0012
   [21]: https://sites.google.com/site/hwachou/publication/Zhou11PSB.pdf?attredirects=0
   [22]: http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btq448?ijkey=FUE0Gg1povcihuH&keytype=ref
   [23]: https://sites.google.com/site/hwachou/publication/ZhouSehlSinsheimerSobelLange10RVLasso.pdf?attredirects=0
   [24]: http://projecteuclid.org/DPubS?verb=Display&version=1.0&service=UI&handle=euclid.ss/1294167962&page=record
   [25]: https://sites.google.com/site/hwachou/publication/ZhouSuchardLange10GPU.pdf?attredirects=0
   [26]: http://onlinelibrary.wiley.com/doi/10.1111/j.1467-9469.2009.00681.x/abstract
   [27]: https://sites.google.com/site/hwachou/publication/ZhouLange10Annealing.pdf?attredirects=0
   [28]: http://pubs.amstat.org/doi/abs/10.1198/jcgs.2010.09014
   [29]: https://sites.google.com/site/hwachou/publication/ZhouLange10MM.pdf?attredirects=0
   [30]: http://cancerres.aacrjournals.org/cgi/content/abstract/69/24/9481
   [31]: https://sites.google.com/site/hwachou/publication/Sehl09CancerRes.pdf?attredirects=0
   [32]: http://www.tandfonline.com/doi/abs/10.1198/tast.2009.08278
   [33]: https://sites.google.com/site/hwachou/publication/ZhouLange2009MovieRating.pdf?attredirects=0
   [34]: http://www.netflixprize.com/index
   [35]: http://www.grouplens.org/node/73
   [36]: http://projecteuclid.org/DPubS?service=UI&version=1.0&verb=Display&handle=euclid.aap/1240319585
   [37]: https://sites.google.com/site/hwachou/publication/ZhouLange09CompositionChains.pdf?attredirects=0
   [38]: http://www.stats.ox.ac.uk/%7Egriff/
   [39]: http://sankhya.isical.ac.in/search/69_2/2007007.html
   [40]: http://projecteuclid.org/DPubS?service=UI&version=1.0&verb=Display&handle=euclid.aoap/1241702249
   [41]: https://sites.google.com/site/hwachou/publication/KhareZhou09AAP.pdf?attredirects=0
   [42]: https://sites.google.com/site/hwachou/publication/rubik.pdf?attredirects=0
